File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/SmallLaserEffect.ts
import { FairyGUI } from "csharp";
import { Mathf } from "../../../../../Core/Module/Math/Mathf";
import Color from "../../DataDefine/Color";
import FileHelper from "../../FileHelper";
import UIHelper from "../../UI/UIHelper";
import { AbstractEffect } from "../AbstractEffect";

export class SmallLaserEffect extends AbstractEffect{
    public Loader1:FairyGUI.GLoader;
    public Loader2:FairyGUI.GLoader;
    private _rotation:number;
    private _scale:number = 1;
    public constructor(private _sX:number,private _sY:number,private _dX:number,private _dY:number) {
        super();
        this.Loader1 = UIHelper.CreateGLoader();
        this.Loader2 = UIHelper.CreateGLoader();
        this.Loader1.url = this.Loader2.url = FileHelper.FormatPath("Vfx/combat/laserThin");
        let dst = Mathf.Distance(_sX,_sY,_dX,_dY);
        this.Loader1.color = this.Loader2.color = Color.Cyan.UnityColor;
        this._rotation = Mathf.Atan2(_dX - _sX,_dY - _sY);
        this._rotation *= Mathf.Rad2Deg;
        this._rotation = -this._rotation + 90;
        this.Loader1.blendMode = this.Loader2.blendMode = FairyGUI.BlendMode.Add;
        this.Loader1.TweenFade(1,0.5).SetEase(FairyGUI.EaseType.QuadOut);
        this.Loader1.TweenFade(0,0.5).SetEase(FairyGUI.EaseType.BounceIn).SetDelay(0.5);
        this.Loader2.TweenFade(1,0.5).SetEase(FairyGUI.EaseType.QuadOut);
        this.Loader2.TweenFade(0,0.5).SetEase(FairyGUI.EaseType.BounceIn).SetDelay(0.5);
        this.Loader1.rotation = this.Loader2.rotation = this._rotation;
    }
    
    Update() {
        super.Update();
        if(this.Duration >= 1) {
            this.IsDone = true;
            this.Loader1.Dispose();
            this.Loader2.Dispose();
            this.Loader2.height = Mathf.Floor(Mathf.RandomRange(50,91));
            this.Loader2.scaleX = this._scale + Mathf.RandomRange(-0.02,0.02);
            this.Loader1.scaleX = this._scale + Mathf.RandomRange(-0.01,0.01);
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/SmallLaserEffect.cs
using Core.Utils;
using FairyGUI;
using ModTheSpire.DataDefine;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class SmallLaserEffect : AbstractEffect
    {
        private float mX;
        private float mY;
        private float mX2;
        private float mY2;
        private GLoader mLoader1;
        private GLoader mLoader2;
        private float mRotation;
        private float mScale = 1;

        public SmallLaserEffect(float x, float y, float x2, float y2)
        {
            mX = x;
            mY = y;
            mX2 = x2;
            mY2 = y2;
            mLoader1 = UIObjectFactory.NewObject(ObjectType.Loader).asLoader;
            mLoader2 = UIObjectFactory.NewObject(ObjectType.Loader).asLoader;
            mLoader1.url = mLoader2.url = "Vfx/combat/laserThin";
            mLoader1.color = mLoader2.color = MTSColor.Cyan.Clone();
            this.mRotation = Mathf.Atan2(x2 - x, y2 - y);
            this.mRotation *= Mathf.Rad2Deg;
            this.mRotation = -this.mRotation + 90;
            mLoader1.blendMode = mLoader2.blendMode = BlendMode.Add;
            mLoader1.TweenFade(1,0.5f).SetEase(EaseType.QuadOut);
            mLoader1.TweenFade(0,0.5f).SetEase(EaseType.BounceIn).SetDelay(0.5f);
            mLoader2.TweenFade(1,0.5f).SetEase(EaseType.QuadOut);
            mLoader2.TweenFade(0,0.5f).SetEase(EaseType.BounceIn).SetDelay(0.5f);
            mLoader1.rotation = mLoader2.rotation = mRotation;
        }

        public override void Update()
        {
            base.Update();
            if(this.Duration >= 1) {
                this.IsDone = true;
                mLoader1.Dispose();
                mLoader2.Dispose();
                mLoader2.height = Mathf.Floor(Random.Range(50,91));
                mLoader2.scaleX = this.mScale + Random.Range(-0.02f,0.02f);
                mLoader1.scaleX = this.mScale + Random.Range(-0.01f,0.01f);
            }
        }
    }
}